provider/pkg/gen/compatibleTokens.go

```
@@ -0,0 +1,84 @@
+	lookup := CompatibleTokensLookup{
+		byLoweredModuleResourceName: map[ModuleLowered]map[ResourceLowered]map[Token]struct{}{},
+		byNormalizedPath:            map[NormalizedPath]map[Token]struct{}{},
+	}
+	if tokenPaths == nil {
+		tokenPaths = map[string]string{}
+	}
+	for token, path := range util.MapOrdered(tokenPaths) {
+		tokenParts := strings.Split(token, ":")
+		if len(tokenParts) != 3 {
+			return nil, fmt.Errorf("malformed token '%s'", token)
+		}
+		tokenModule, resourceName := tokenParts[1], tokenParts[2]
+		moduleName, _, _ := strings.Cut(tokenModule, "/")
+		lookup.add(openapi.ModuleName(moduleName), resourceName, path, token)
+	}
+	return &lookup, nil
+}
+
+func (lookup *CompatibleTokensLookup) add(moduleName openapi.ModuleName, resourceName openapi.ResourceName, path string, token string) {
+	normalizedPath := NormalizedPath(paths.NormalizePath(path))
+	moduleLowered := ModuleLowered(moduleName.Lowered())
+	resourceLowered := ResourceLowered(strings.ToLower(resourceName))
+	typedToken := Token(token)
+
+	if _, ok := lookup.byLoweredModuleResourceName[moduleLowered]; !ok {
+		lookup.byLoweredModuleResourceName[moduleLowered] = map[ResourceLowered]map[Token]struct{}{}
+	}
+	if _, ok := lookup.byLoweredModuleResourceName[moduleLowered][resourceLowered]; !ok {
+		lookup.byLoweredModuleResourceName[moduleLowered][resourceLowered] = map[Token]struct{}{}
+	}
+
+	lookup.byLoweredModuleResourceName[moduleLowered][resourceLowered][typedToken] = struct{}{}
+	if _, ok := lookup.byNormalizedPath[normalizedPath]; !ok {
+		lookup.byNormalizedPath[normalizedPath] = map[Token]struct{}{}
+	}
+	lookup.byNormalizedPath[normalizedPath][typedToken] = struct{}{}
+}
+
+func (lookup *CompatibleTokensLookup) FindCompatibleTokens(moduleName openapi.ModuleName, resourceName openapi.ResourceName, path string) []string {
+	if lookup == nil || lookup.byLoweredModuleResourceName == nil || lookup.byNormalizedPath == nil {
+		return nil
```

Silently returning an empty result could lead to surprising behavior. Return `([]string, error)`?

---

Yeah, I did have a separate method `IsPopulated` to assert there's something in the collection, but then we'd still just carry on for now even if there's nothing in here because it's not populated for v2.

Perhaps I add that check back in and fail hard if it's not populated and the major version is v3?